cloud/vsphere/machineactuator.go

```diff
@@ -710,6 +712,17 @@ func (vc *VsphereClient) validateCluster(cluster *clusterv1.Cluster) error {
 	return nil
 }
 
+func getKubeadmToken() (string, error) {
```

Comment:
This seems to be exactly what's in the gce actuator. Why not put this kubeadm.go and call that directly?

Reply:
I appreciate the sentiment to avoid copy & paste and semi agree. While it is the same, in terms of code complexity and layering I would vote against putting it in kubeadm.go. Kubeadm.go, right now, only deals with calling kubeadm, passing parameters correcdtly, and handling errors. A programmer can think of it as an easier way to call kubeadm. Something  calling kubeadm with specific parameters like this should probably be in another layer sitting on top of kubeadm.go if we are to try to cut down the duplication. 

In addition, we are mocking the kubeadm implementation in GCP's machine actuator (and testing the output). My inclination is to put it in another tested layer or simply leave the code as divergent. Happy to create something like a Token fetcher or the like that can go within the Kubeadm package if we think that is the right thing do.

Reply:
Sure. My concern is more with the error handling and string trim spaces. `TokenCreate` should return either an error or a complete token that the callee can just use. I don't know if errors are logged at the caller or callee level in k8s/go.

Reply:
The idea is to make kubeadm.TokenCreate(...) the exact same thing as typing out "kubeadm token create" on the command line. Note that there are flags that make the token create command print more than simply a single token. I'll add some comments to kubeadm.go to make it clear that to limit complexity that layer is merely responsible for executing the commands directly. Created an issue for that for a follow up PR here: https://github.com/kubernetes-sigs/cluster-api/issues/332.

TBH the string trimming is likely unnecessary and could be removed. It was a bit of defensive programminmg but probably a bad idea because it creates confusion / uncertainty about what kubeadm does. Errors should be logged at a higher level than the kubeadm package as it can and will be used by multiple different controllers / other modules and each will have their own logging requirements and contexts for why they are calling execing kubeadm. 

As far as removing the code duplication between the gce machine actuator, there's a bit of a difference in that the kubeadm implementation is being injected into the GCE machine actuator for unit tests, etc, plus it seems likely that different providers will want to have different lifecycles for their tokens (for example on some 10 minutes may not be long enough for a machine to join the cluster, etc). GCP actually goes through an interface "GCEClientKubeadm" to use kubeadm. 

I think there is a common code location that should be created for these but it seems like something that would be in a common lib place for controller code. Maybe under pkg/controller in a package in a kubeadm package there. I'd like to keep the scope of this PR to adding this vsphere functionality, so I created an issue for removing this duplication: https://github.com/kubernetes-sigs/cluster-api/issues/333.

Reply:
Considering other providers will be copying this method verbatim, and if the TrimSpace is not needed, please remove it.

Otherwise lgtm.

Reply:
Thanks, I just realized the reason why it is needed is because the kubeadm command is appending a newline to the end of the create token output -- see the tests in the GCP version of  machine_actuator.

I updated the follow up code duplication issue, https://github.com/kubernetes-sigs/cluster-api/issues/333, with this information.